File: app/src/main/java/com/github/premnirmal/ticker/settings/SettingsActivity.java
package com.github.premnirmal.ticker.settings;

import android.app.AlertDialog;
import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.preference.CheckBoxPreference;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceActivity;
import android.support.v7.widget.Toolbar;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.github.premnirmal.ticker.CrashLogger;
import com.devpaul.filepickerlibrary.FilePickerActivity;
import com.github.premnirmal.ticker.Analytics;
import com.github.premnirmal.ticker.Injector;
import com.github.premnirmal.ticker.Tools;
import com.github.premnirmal.ticker.model.IStocksProvider;
import com.github.premnirmal.ticker.widget.StockWidget;
import com.github.premnirmal.tickerwidget.BuildConfig;
import com.github.premnirmal.tickerwidget.R;

import java.io.File;

import javax.inject.Inject;

import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper;
import uk.co.chrisjenx.calligraphy.CalligraphyTypefaceSpan;
import uk.co.chrisjenx.calligraphy.TypefaceUtils;

/**
 * Created by premnirmal on 01/09/15.
 */
public class SettingsActivity extends PreferenceActivity {

    @Inject
    IStocksProvider stocksProvider;

    @Inject
    SharedPreferences preferences;

    private class DefaultPreferenceChangeListener implements Preference.OnPreferenceChangeListener {

        @Override
        public boolean onPreferenceChange(Preference preference, Object newValue) {

            return false;
        }
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Injector.inject(this);
        setContentView(R.layout.activity_preferences);
    }

    @Override
    protected void attachBaseContext(Context newBase) {
        super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase));
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        final Toolbar bar = (Toolbar) findViewById(R.id.toolbar);
        bar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });

        getListView().addFooterView(LayoutInflater.from(this).inflate(R.layout.preferences_footer, null, false));

        final TextView versionView = (TextView) findViewById(R.id.version);
        final SpannableStringBuilder sBuilder = new SpannableStringBuilder();
        sBuilder.append("Version " + BuildConfig.VERSION_NAME);
        final CalligraphyTypefaceSpan typefaceSpan = new CalligraphyTypefaceSpan(TypefaceUtils.load(getAssets(), "fonts/alegreya-black-italic.ttf"));
        sBuilder.setSpan(typefaceSpan, 0, sBuilder.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        versionView.setText(sBuilder);
        setupSimplePreferencesScreen();
    }

    /**
     * Shows the simplified settings UI if the device configuration if the
     * device configuration dictates that a simplified, single-pane UI should be
     * shown.
     */
    private void setupSimplePreferencesScreen() {
        // In the simplified UI, fragments are not used at all and we instead
        // use the older PreferenceActivity APIs.

        // Add 'general' preferences.
        addPreferencesFromResource(R.xml.prefs);

        {
            final Preference exportPref = findPreference("EXPORT");
            exportPref.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    new FileExportTask() {
                        @Override
                        protected void onPostExecute(String result) {
                            if (result == null) {
                                showDialog(getString(R.string.error_exporting));
                                CrashLogger.logException(new Throwable("Error exporting tickers"));
                            } else {
                               showDialog("Exported to " + result);
                            }
                        }
                    }.execute(stocksProvider.getTickers().toArray());
                    return true;
                }
            });
        }

        {
            final Preference sharePref = findPreference("SHARE");
            sharePref.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    Analytics.trackSettingsChange("SHARE", TextUtils.join(",", stocksProvider.getTickers().toArray()));
                    final File file = Tools.getTickersFile();
                    if(file.exists()) {
                        shareTickers();
                    } else {
                        new FileExportTask() {
                            @Override
                            protected void onPostExecute(String result) {
                                if (result == null) {
                                    showDialog(getString(R.string.error_sharing));
                                    CrashLogger.logException(new Throwable("Error sharing tickers"));
                                } else {
                                    shareTickers();
                                }
                            }
                        }.execute(stocksProvider.getTickers().toArray());
                    }
                    return true;
                }
            });
        }

        {
            final Preference importPref = findPreference("IMPORT");
            importPref.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    final Intent filePickerIntent = new Intent(SettingsActivity.this, FilePickerActivity.class);
                    filePickerIntent.putExtra(FilePickerActivity.REQUEST_CODE, FilePickerActivity.REQUEST_FILE);
                    filePickerIntent.putExtra(FilePickerActivity.INTENT_EXTRA_COLOR_ID, R.color.color_primary);
                    startActivityForResult(filePickerIntent, FilePickerActivity.REQUEST_FILE);
                    return true;
                }
            });
        }

        {
            final ListPreference fontSizePreference = (ListPreference) findPreference(Tools.FONT_SIZE);
            final int size = preferences.getInt(Tools.FONT_SIZE, 1);
            fontSizePreference.setValueIndex(size);
            fontSizePreference.setSummary(fontSizePreference.getEntries()[size]);
            fontSizePreference.setOnPreferenceChangeListener(new DefaultPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object value) {
                    super.onPreferenceChange(preference,value);
                    final String stringValue = value.toString();
                    final ListPreference listPreference = (ListPreference) preference;
                    final int index = listPreference.findIndexOfValue(stringValue);
                    preferences.edit().remove(Tools.FONT_SIZE).putInt(Tools.FONT_SIZE, index).apply();
                    broadcastUpdateWidget();
                    fontSizePreference.setSummary(fontSizePreference.getEntries()[index]);
                    Toast.makeText(SettingsActivity.this, R.string.text_size_updated_message, Toast.LENGTH_SHORT).show();
                    return true;
                }
            });
        }

        {
            final ListPreference bgPreference = (ListPreference) findPreference(Tools.WIDGET_BG);
            final int bgIndex = preferences.getInt(Tools.WIDGET_BG, 0);
            bgPreference.setValueIndex(bgIndex);
            bgPreference.setSummary(bgPreference.getEntries()[bgIndex]);
            bgPreference.setOnPreferenceChangeListener(new DefaultPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object value) {
                    super.onPreferenceChange(preference,value);
                    final String stringValue = value.toString();
                    final ListPreference listPreference = (ListPreference) preference;
                    final int index = listPreference.findIndexOfValue(stringValue);
                    preferences.edit().putInt(Tools.WIDGET_BG, index).apply();
                    broadcastUpdateWidget();
                    bgPreference.setSummary(bgPreference.getEntries()[index]);
                    Toast.makeText(SettingsActivity.this, R.string.bg_updated_message, Toast.LENGTH_SHORT).show();
                    return true;
                }
            });
        }

        {
            final ListPreference layoutTypePref = (ListPreference) findPreference(Tools.LAYOUT_TYPE);
            final int typeIndex = preferences.getInt(Tools.LAYOUT_TYPE, 0);
            layoutTypePref.setValueIndex(typeIndex);
            layoutTypePref.setSummary(layoutTypePref.getEntries()[typeIndex]);
            layoutTypePref.setOnPreferenceChangeListener(new DefaultPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object value) {
                    super.onPreferenceChange(preference,value);
                    final String stringValue = value.toString();
                    final ListPreference listPreference = (ListPreference) preference;
                    final int index = listPreference.findIndexOfValue(stringValue);
                    preferences.edit().putInt(Tools.LAYOUT_TYPE, index).apply();
                    broadcastUpdateWidget();
                    layoutTypePref.setSummary(layoutTypePref.getEntries()[index]);
                    Toast.makeText(SettingsActivity.this, R.string.layout_updated_message, Toast.LENGTH_SHORT).show();
                    return true;
                }
            });
        }

        {
            final ListPreference textColorPreference = (ListPreference) findPreference(Tools.TEXT_COLOR);
            final int colorIndex = preferences.getInt(Tools.TEXT_COLOR, 0);
            textColorPreference.setValueIndex(colorIndex);
            textColorPreference.setSummary(textColorPreference.getEntries()[colorIndex]);
            textColorPreference.setOnPreferenceChangeListener(new DefaultPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object value) {
                    super.onPreferenceChange(preference,value);
                    final String stringValue = value.toString();
                    final ListPreference listPreference = (ListPreference) preference;
                    final int index = listPreference.findIndexOfValue(stringValue);
                    preferences.edit().putInt(Tools.TEXT_COLOR, index).apply();
                    broadcastUpdateWidget();
                    CharSequence color = textColorPreference.getEntries()[index];
                    textColorPreference.setSummary(color);
                    Toast.makeText(SettingsActivity.this, R.string.text_coor_updated_message, Toast.LENGTH_SHORT).show();
                    return true;
                }
            });
        }

        {
            final ListPreference refreshPreference = (ListPreference) findPreference(Tools.UPDATE_INTERVAL);
            final int refreshIndex = preferences.getInt(Tools.UPDATE_INTERVAL, 1);
            refreshPreference.setValueIndex(refreshIndex);
            refreshPreference.setSummary(refreshPreference.getEntries()[refreshIndex]);
            refreshPreference.setOnPreferenceChangeListener(new DefaultPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object value) {
                    super.onPreferenceChange(preference,value);
                    final String stringValue = value.toString();
                    final ListPreference listPreference = (ListPreference) preference;
                    final int index = listPreference.findIndexOfValue(stringValue);
                    preferences.edit().putInt(Tools.UPDATE_INTERVAL, index).apply();
                    broadcastUpdateWidget();
                    refreshPreference.setSummary(refreshPreference.getEntries()[index]);
                    Toast.makeText(SettingsActivity.this, R.string.refresh_updated_message, Toast.LENGTH_SHORT).show();
                    return true;
                }
            });
        }

        {
            final CheckBoxPreference autoSortPreference = (CheckBoxPreference) findPreference(Tools.SETTING_AUTOSORT);
            final boolean autoSort = Tools.autoSortEnabled();
            autoSortPreference.setChecked(autoSort);
            autoSortPreference.setOnPreferenceChangeListener(new DefaultPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object value) {
                    super.onPreferenceChange(preference,value);
                    final boolean checked = (boolean) value;
                    preferences.edit().putBoolean(Tools.SETTING_AUTOSORT, checked).apply();
                    return true;
                }
            });
        }

        {
            final CheckBoxPreference boldChangePreference = (CheckBoxPreference) findPreference(Tools.BOLD_CHANGE);
            final boolean bold = Tools.boldEnabled();
            boldChangePreference.setChecked(bold);
            boldChangePreference.setOnPreferenceChangeListener(new DefaultPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object value) {
                    super.onPreferenceChange(preference,value);
                    final boolean checked = (boolean) value;
                    preferences.edit().putBoolean(Tools.BOLD_CHANGE, checked).apply();
                    return true;
                }
            });
        }

        {
            final TimePreference startTimePref = (TimePreference) findPreference(Tools.START_TIME);
            startTimePref.setSummary(preferences.getString(Tools.START_TIME, "09:30"));
            startTimePref.setOnPreferenceChangeListener(new DefaultPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object newValue) {
                    super.onPreferenceChange(preference,newValue);
                    preferences.edit().putString(Tools.START_TIME, newValue.toString()).apply();
                    startTimePref.setSummary(newValue.toString());
                    Toast.makeText(SettingsActivity.this, R.string.start_time_updated, Toast.LENGTH_SHORT).show();
                    return true;
                }
            });
        }

        {
            final TimePreference endTimePref = (TimePreference) findPreference(Tools.END_TIME);
            endTimePref.setSummary(preferences.getString(Tools.END_TIME, "16:30"));
            endTimePref.setOnPreferenceChangeListener(new DefaultPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object newValue) {
                    super.onPreferenceChange(preference,newValue);
                    preferences.edit().putString(Tools.END_TIME, newValue.toString()).apply();
                    endTimePref.setSummary(newValue.toString());
                    Toast.makeText(SettingsActivity.this, R.string.end_time_updated, Toast.LENGTH_SHORT).show();
                    return true;
                }
            });
        }
    }

    private void shareTickers() {
        final Intent intent = new Intent(Intent.ACTION_SEND);
        intent.setType("text/plain");
        intent.putExtra(Intent.EXTRA_EMAIL, new String[]{});
        intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.my_stock_portfolio));
        intent.putExtra(Intent.EXTRA_TEXT, getString(R.string.share_email_subject));
        final File file = Tools.getTickersFile();
        if (!file.exists() || !file.canRead()) {
            showDialog(getString(R.string.error_sharing));
            CrashLogger.logException(new Throwable("Error sharing tickers"));
            return;
        }
        final Uri uri = Uri.fromFile(file);
        intent.putExtra(Intent.EXTRA_STREAM, uri);
        startActivity(Intent.createChooser(intent, getString(R.string.action_share)));
    }

    private void broadcastUpdateWidget() {
        final Intent intent = new Intent(getApplicationContext(), StockWidget.class);
        intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
        final AppWidgetManager widgetManager = AppWidgetManager.getInstance(getApplicationContext());
        final int[] ids = widgetManager.getAppWidgetIds(new ComponentName(getApplicationContext(), StockWidget.class));
        widgetManager.notifyAppWidgetViewDataChanged(ids, android.R.id.list);
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids);
        sendBroadcast(intent);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == FilePickerActivity.REQUEST_FILE
                && resultCode == RESULT_OK) {
            final String filePath = data.
                    getStringExtra(FilePickerActivity.FILE_EXTRA_DATA_PATH);
            if (filePath != null) {
                new FileImportTask(stocksProvider) {
                    @Override
                    protected void onPostExecute(Boolean result) {
                        if (result) {
                            showDialog(getString(R.string.ticker_import_success));
                        } else {
                            showDialog(getString(R.string.ticker_import_fail));
                        }
                    }
                }.execute(filePath);
            }
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    private void showDialog(String message) {
        new AlertDialog.Builder(this)
                .setMessage(message)
                .setNeutralButton("OK", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                })
                .show();
    }
}

File: app/src/main/kotlin/com/github/premnirmal/ticker/settings/SettingsActivity.kt
package com.github.premnirmal.ticker.settings

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.DialogInterface.OnClickListener
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.preference.ListPreference
import android.preference.Preference
import android.preference.PreferenceActivity
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v4.content.FileProvider
import android.view.LayoutInflater
import com.github.premnirmal.ticker.AppPreferences
import com.github.premnirmal.ticker.base.BaseActivity.Companion.getStatusBarHeight
import com.github.premnirmal.ticker.base.BaseActivity.Companion.showDialog
import com.github.premnirmal.ticker.components.InAppMessage
import com.github.premnirmal.ticker.components.Injector
import com.github.premnirmal.ticker.model.IStocksProvider
import com.github.premnirmal.ticker.widget.WidgetDataProvider
import com.github.premnirmal.tickerwidget.BuildConfig
import com.github.premnirmal.tickerwidget.R
import com.nbsp.materialfilepicker.MaterialFilePicker
import com.nbsp.materialfilepicker.ui.FilePickerActivity
import kotlinx.android.synthetic.main.activity_preferences.toolbar
import kotlinx.android.synthetic.main.preferences_footer.version
import timber.log.Timber
import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper
import java.io.File
import java.util.regex.Pattern
import javax.inject.Inject

/**
 * Created by premnirmal on 2/27/16.
 */
class SettingsActivity : PreferenceActivity(), ActivityCompat.OnRequestPermissionsResultCallback {

  companion object {
    private val REQCODE_WRITE_EXTERNAL_STORAGE = 850
    private val REQCODE_READ_EXTERNAL_STORAGE = 851
    private val REQCODE_WRITE_EXTERNAL_STORAGE_SHARE = 852
    private val REQCODE_FILE_WRITE = 853
  }

  @Inject internal lateinit var stocksProvider: IStocksProvider
  @Inject internal lateinit var widgetDataProvider: WidgetDataProvider
  @Inject internal lateinit var preferences: SharedPreferences
  @Inject internal lateinit var appPreferences: AppPreferences

  override fun onPause() {
    super.onPause()
    broadcastUpdateWidget()
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    Injector.appComponent.inject(this)
    setContentView(R.layout.activity_preferences)
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
      toolbar.setPadding(toolbar.paddingLeft, getStatusBarHeight(),
          toolbar.paddingRight, toolbar.paddingBottom)
    }
  }

  override fun attachBaseContext(newBase: Context) {
    super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase))
  }

  override fun onPostCreate(savedInstanceState: Bundle?) {
    super.onPostCreate(savedInstanceState)
    val bar = toolbar
    bar.setNavigationOnClickListener({ onBackPressed() })

    listView.addFooterView(
        LayoutInflater.from(this).inflate(R.layout.preferences_footer, null, false))

    val versionView = version
    val vName = "v${BuildConfig.VERSION_NAME}"
    versionView.text = vName
    setupSimplePreferencesScreen()
  }

  /**
   * Shows the simplified settings UI if the device configuration if the
   * device configuration dictates that a simplified, single-pane UI should be
   * shown.
   */
  @SuppressLint("CommitPrefEdits")
  private fun setupSimplePreferencesScreen() {
    // In the simplified UI, fragments are not used at all and we instead
    // use the older PreferenceActivity APIs.

    // Add 'general' preferences.
    addPreferencesFromResource(R.xml.prefs)

    run({
      val nukePref = findPreference(AppPreferences.SETTING_NUKE)
      nukePref.onPreferenceClickListener = Preference.OnPreferenceClickListener {
        showDialog(getString(R.string.are_you_sure), OnClickListener { _, _ ->
          val hasUserAlreadyRated = appPreferences.hasUserAlreadyRated()
          Timber.w(RuntimeException("Nuked from settings!"))
          preferences.edit().clear().commit()
          val directory = filesDir.path + "$packageName/shared_prefs/"
          val sharedPreferenceFile = File(directory)
          val listFiles = sharedPreferenceFile.listFiles()
          listFiles?.forEach { file ->
            file?.delete()
          }
          preferences.edit().putBoolean(AppPreferences.DID_RATE, hasUserAlreadyRated).commit()
          System.exit(0)
        })
        true
      }
    })

    run({
      val exportPref = findPreference(AppPreferences.SETTING_EXPORT)
      exportPref.onPreferenceClickListener = Preference.OnPreferenceClickListener {
        if (needsPermissionGrant()) {
          askForExternalStoragePermissions(REQCODE_WRITE_EXTERNAL_STORAGE)
        } else {
          exportTickers()
        }
        true
      }
    })

    run({
      val sharePref = findPreference(AppPreferences.SETTING_SHARE)
      sharePref.onPreferenceClickListener = Preference.OnPreferenceClickListener {
        if (needsPermissionGrant()) {
          askForExternalStoragePermissions(REQCODE_WRITE_EXTERNAL_STORAGE_SHARE)
        } else {
          exportAndShareTickers()
        }
        true
      }
    })

    run({
      val importPref = findPreference(AppPreferences.SETTING_IMPORT)
      importPref.onPreferenceClickListener = Preference.OnPreferenceClickListener {
        if (needsPermissionGrant()) {
          askForExternalStoragePermissions(REQCODE_READ_EXTERNAL_STORAGE)
        } else {
          launchImportIntent()
        }
        true
      }
    })

    run({
      val fontSizePreference = findPreference(AppPreferences.FONT_SIZE) as ListPreference
      val size = preferences.getInt(AppPreferences.FONT_SIZE, 1)
      fontSizePreference.setValueIndex(size)
      fontSizePreference.summary = fontSizePreference.entries[size]
      fontSizePreference.onPreferenceChangeListener = object : DefaultPreferenceChangeListener() {
        override fun onPreferenceChange(preference: Preference, newValue: Any): Boolean {
          val stringValue = newValue.toString()
          val listPreference = preference as ListPreference
          val index = listPreference.findIndexOfValue(stringValue)
          preferences.edit().remove(AppPreferences.FONT_SIZE).putInt(AppPreferences.FONT_SIZE,
              index).apply()
          broadcastUpdateWidget()
          fontSizePreference.summary = fontSizePreference.entries[index]
          InAppMessage.showMessage(this@SettingsActivity, R.string.text_size_updated_message)
          return true
        }
      }
    })

    run({
      val refreshPreference = findPreference(AppPreferences.UPDATE_INTERVAL) as ListPreference
      val refreshIndex = preferences.getInt(AppPreferences.UPDATE_INTERVAL, 1)
      refreshPreference.setValueIndex(refreshIndex)
      refreshPreference.summary = refreshPreference.entries[refreshIndex]
      refreshPreference.onPreferenceChangeListener = object : DefaultPreferenceChangeListener() {
        override fun onPreferenceChange(preference: Preference, newValue: Any): Boolean {
          val stringValue = newValue.toString()
          val listPreference = preference as ListPreference
          val index = listPreference.findIndexOfValue(stringValue)
          preferences.edit().putInt(AppPreferences.UPDATE_INTERVAL, index).apply()
          broadcastUpdateWidget()
          refreshPreference.summary = refreshPreference.entries[index]
          InAppMessage.showMessage(this@SettingsActivity, R.string.refresh_updated_message)
          return true
        }
      }
    })

    run({
      val startTimePref = findPreference(AppPreferences.START_TIME) as TimePreference
      startTimePref.summary = preferences.getString(AppPreferences.START_TIME, "09:30")
      startTimePref.onPreferenceChangeListener = object : DefaultPreferenceChangeListener() {
        override fun onPreferenceChange(preference: Preference, newValue: Any): Boolean {
          val startTimez = appPreferences.timeAsIntArray(newValue.toString())
          val endTimez = appPreferences.endTime()
          if (endTimez[0] == startTimez[0] && endTimez[1] == startTimez[1]) {
            showDialog(getString(R.string.incorrect_time_update_error))
            return false
          } else {
            preferences.edit().putString(AppPreferences.START_TIME, newValue.toString()).apply()
            startTimePref.summary = newValue.toString()
            stocksProvider.schedule()
            InAppMessage.showMessage(this@SettingsActivity, R.string.start_time_updated)
            return true
          }
        }
      }
    })

    run({
      val endTimePref = findPreference(AppPreferences.END_TIME) as TimePreference
      endTimePref.summary = preferences.getString(AppPreferences.END_TIME, "16:30")
      run({
        val endTimez = appPreferences.endTime()
        val startTimez = appPreferences.startTime()
        if (endTimez[0] == startTimez[0] && endTimez[1] == startTimez[1]) {
          endTimePref.setSummary(R.string.incorrect_time_update_error)
        }
      })
      endTimePref.onPreferenceChangeListener = object : DefaultPreferenceChangeListener() {
        override fun onPreferenceChange(preference: Preference, newValue: Any): Boolean {
          val endTimez = appPreferences.timeAsIntArray(newValue.toString())
          val startTimez = appPreferences.startTime()
          if (endTimez[0] == startTimez[0] && endTimez[1] == startTimez[1]) {
            showDialog(getString(R.string.incorrect_time_update_error))
            return false
          } else {
            preferences.edit().putString(AppPreferences.END_TIME, newValue.toString()).apply()
            endTimePref.summary = newValue.toString()
            stocksProvider.schedule()
            InAppMessage.showMessage(this@SettingsActivity, R.string.end_time_updated)
            return true
          }
        }
      }
    })
  }

  private fun needsPermissionGrant(): Boolean {
    return Build.VERSION.SDK_INT >= 23 &&
        ContextCompat.checkSelfPermission(this@SettingsActivity,
            Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED
  }

  private fun askForExternalStoragePermissions(reqCode: Int) {
    ActivityCompat.requestPermissions(this@SettingsActivity,
        arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.READ_EXTERNAL_STORAGE),
        reqCode)
  }

  private fun exportAndShareTickers() {
    val file = AppPreferences.tickersFile
    if (file.exists()) {
      shareTickers()
    } else {
      object : FileExportTask() {
        override fun onPostExecute(result: String?) {
          if (result == null) {
            showDialog(getString(R.string.error_sharing))
            Timber.w(Throwable("Error sharing tickers"))
          } else {
            shareTickers()
          }
        }
      }.execute(stocksProvider.getTickers())
    }
  }

  private fun launchImportIntent() {
    MaterialFilePicker()
        .withActivity(this)
        .withRequestCode(REQCODE_FILE_WRITE)
        .withFilter(Pattern.compile(
            ".*\\.txt$")) // Filtering files and directories by file name using regexp
        .start()
  }

  private fun exportTickers() {
    object : FileExportTask() {
      override fun onPostExecute(result: String?) {
        if (result == null) {
          showDialog(getString(R.string.error_exporting))
          Timber.w(Throwable("Error exporting tickers"))
        } else {
          showDialog(getString(R.string.exported_to, result))
        }
      }
    }.execute(stocksProvider.getTickers())
  }

  private fun shareTickers() {
    val intent = Intent(Intent.ACTION_SEND)
    intent.type = "text/plain"
    intent.putExtra(Intent.EXTRA_EMAIL, arrayOf<String>())
    intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.my_stock_portfolio))
    intent.putExtra(Intent.EXTRA_TEXT, getString(R.string.share_email_subject))
    val file = AppPreferences.tickersFile
    if (!file.exists() || !file.canRead()) {
      showDialog(getString(R.string.error_sharing))
      Timber.w(Throwable("Error sharing tickers"))
      return
    }
    val uri = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
      FileProvider.getUriForFile(this, BuildConfig.APPLICATION_ID + ".provider", file)
    } else {
      Uri.fromFile(file)
    }
    intent.putExtra(Intent.EXTRA_STREAM, uri)
    val launchIntent = Intent.createChooser(intent, getString(R.string.action_share))
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
      launchIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
    }
    startActivity(launchIntent)
  }

  private fun broadcastUpdateWidget() {
    widgetDataProvider.broadcastUpdateAllWidgets()
  }

  override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
      grantResults: IntArray) {
    when (requestCode) {
      REQCODE_WRITE_EXTERNAL_STORAGE -> {
        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
          exportTickers()
        } else {
          showDialog(getString(R.string.cannot_export_msg))
        }
      }
      REQCODE_READ_EXTERNAL_STORAGE -> {
        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
          launchImportIntent()
        } else {
          showDialog(getString(R.string.cannot_import_msg))
        }
      }
      REQCODE_WRITE_EXTERNAL_STORAGE_SHARE -> {
        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
          exportAndShareTickers()
        } else {
          showDialog(getString(R.string.cannot_share_msg))
        }
      }
    }
  }

  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
    if (requestCode == REQCODE_FILE_WRITE && resultCode == Activity.RESULT_OK) {
      val filePath = data?.getStringExtra(FilePickerActivity.RESULT_FILE_PATH)
      if (filePath != null) {
        object : FileImportTask(widgetDataProvider) {
          override fun onPostExecute(result: Boolean?) {
            if (result != null && result) {
              showDialog(getString(R.string.ticker_import_success))
            } else {
              showDialog(getString(R.string.ticker_import_fail))
            }
          }
        }.execute(filePath)
      }
    }
    super.onActivityResult(requestCode, resultCode, data)
  }
}